src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,744 @@
+        {
+            // Process any bytes remaining in the partial buffer
+            if (_partialBlockBytes > 0)
+            {
+                // Encrypt the partial block (ProcessBlock handles partials by XORing only available bytes)
+                ProcessBlock(_partialBlock, 0, _partialBlockBytes);
+
+                if (isAsync)
+                {
+                    await _baseStream.WriteAsync(_partialBlock.AsMemory(0, _partialBlockBytes), cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    _baseStream.Write(_partialBlock, 0, _partialBlockBytes);
+                }
+
+                _partialBlockBytes = 0;
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (_disposed) return;
+
+            if (disposing)
+            {
+                try
+                {
+                    if (_encrypting && !_authCodeValidated && _headerWritten)
+                    {
+                        // 1. Encrypt remaining partial data
+                        FinalizeEncryptionAsync(false, CancellationToken.None).GetAwaiter().GetResult();
+
+                        // 2. Write Auth Code
+                        WriteAuthCode();
+
+                        if (_baseStream.CanWrite) _baseStream.Flush();
+                    }
+                    else if (!_encrypting && !_authCodeValidated && _headerRead)
+                    {
+                        ValidateAuthCodeCoreAsync(false, CancellationToken.None).GetAwaiter().GetResult();
+                    }
+                }
+                finally
+                {
+                    _aesEncryptor?.Dispose();
+                    _aes.Dispose();
+                    _hmac.Dispose();
+                    // Removed _encryptionBuffer.Dispose()
+
+                    if (!_leaveOpen) _baseStream.Dispose();
+                }
+            }
+
+            _disposed = true;
+            base.Dispose(disposing);
+        }
+        public override async ValueTask DisposeAsync()
+        {
+            if (_disposed) return;
+
+            try
+            {
+                if (_encrypting && !_authCodeValidated && _headerWritten)
+                {
+                    await _baseStream.FlushAsync().ConfigureAwait(false);
+
+                    // 1. Encrypt remaining partial data
+                    await FinalizeEncryptionAsync(true, CancellationToken.None).ConfigureAwait(false);
+
+                    // 2. Write Auth Code
+                    await WriteAuthCodeAsync(CancellationToken.None).ConfigureAwait(false);
+
+                    if (_baseStream.CanWrite) await _baseStream.FlushAsync().ConfigureAwait(false);
+                }
+                else if (!_encrypting && !_authCodeValidated && _headerRead)
+                {
+                    await ValidateAuthCodeCoreAsync(true, CancellationToken.None).ConfigureAwait(false);
```

Comment:
We should validate the auth code immediately as we read the last segment of the data so that we throw during Read(Async) and not during Dispose for corrupted input